Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm

```diff
@@ -0,0 +1,321 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "CoreIPCNSURLRequest.h"
+
+#import "GeneratedSerializers.h"
+#import <wtf/cocoa/VectorCocoa.h>
+
+#if PLATFORM(COCOA) && HAVE(WK_SECURE_CODING_NSURLREQUEST)
+
+@interface NSURLRequest (WKSecureCoding)
+- (NSDictionary *)_webKitPropertyListData;
+- (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;
+@end
+
+namespace WebKit {
+
+#define SET_NSURLREQUESTDATA(NAME, CLASS, WRAPPER)  \
+    id NAME = dict[@#NAME];                 \
+    if ([NAME isKindOfClass:CLASS.class]) { \
+        auto var = WRAPPER(NAME);           \
+        m_data.NAME = WTFMove(var);         \
+    }
+
+#define SET_NSURLREQUESTDATA_PRIMITIVE(NAME, CLASS, PRIMITIVE) \
+    id NAME = dict[@#NAME];                                    \
+    if ([NAME isKindOfClass:CLASS.class])                      \
+        m_data.NAME = [NAME PRIMITIVE##Value];
+
+#define SET_DICT_FROM_OPTIONAL_MEMBER(KEY)                     \
+    if (m_data.KEY)                                            \
+        [dict setObject:m_data.KEY->toID().get() forKey:@#KEY]
+
+#define SET_DICT_FROM_MEMBER(KEY) \
+    [dict setObject:m_data.KEY.toID().get() forKey:@#KEY]
+
+#define SET_DICT_FROM_OPTIONAL_PRIMITIVE(KEY, CLASS, PRIMITIVE)                    \
+    if (m_data.KEY)                                                                \
+        [dict setObject:[NSNumber numberWith##PRIMITIVE:*m_data.KEY] forKey:@#KEY]
+
+#define SET_DICT_FROM_PRIMITIVE(KEY, CLASS, PRIMITIVE) \
+    [dict setObject:[NSNumber numberWith##PRIMITIVE:m_data.KEY] forKey:@#KEY]
+
+CoreIPCNSURLRequest::CoreIPCNSURLRequest(NSURLRequest *request)
+{
+    auto dict = [request _webKitPropertyListData];
+
+    SET_NSURLREQUESTDATA(protocolProperties, NSDictionary, CoreIPCPlistDictionary);
+    SET_NSURLREQUESTDATA_PRIMITIVE(isMutable, NSNumber, bool);
+    SET_NSURLREQUESTDATA(URL, NSURL, CoreIPCURL);
+    SET_NSURLREQUESTDATA_PRIMITIVE(timeout, NSNumber, double);
+
+    id cachePolicy = dict[@"cachePolicy"];
+    if (cachePolicy != nil && [cachePolicy isKindOfClass:[NSNumber class]]) {
```

Comment:
nil check unneeded because isKindOfClass returns false when called on nil.  Several times in this function.